hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/snapshot/filter/TestReclaimableRenameEntryFilter.java

```diff
@@ -0,0 +1,204 @@
+      throws IOException, RocksDBException {
+    setup(1, actualNumberOfSnapshots, index, 4, 2,
+        BucketLayout.OBJECT_STORE);
+    String volume = getVolumes().get(3);
+    String bucket = getBuckets().get(1);
+    index = Math.min(index, actualNumberOfSnapshots);
+    String value = UUID.randomUUID().toString();
+    Table<String, OmKeyInfo> keyTable = getMockedTable(Collections.emptyMap());
+    Table<String, OmDirectoryInfo> directoryTable = getFailingMockedTable();
+    testReclaimableRenameEntryFilter(volume, bucket, index, value, keyTable, directoryTable, true);
+  }
+
+  @ParameterizedTest
+  @MethodSource("testReclaimableFilterArguments")
+  public void testReclaimableRenameEntryWithFSO(int actualNumberOfSnapshots, int index)
+      throws IOException, RocksDBException {
+    setup(1, actualNumberOfSnapshots, index, 4, 2,
+        BucketLayout.FILE_SYSTEM_OPTIMIZED);
+    String volume = getVolumes().get(3);
+    String bucket = getBuckets().get(1);
+    index = Math.min(index, actualNumberOfSnapshots);
+    String value = UUID.randomUUID().toString();
+    Table<String, OmKeyInfo> keyTable = getMockedTable(Collections.emptyMap());
+    Table<String, OmDirectoryInfo> directoryTable = getMockedTable(Collections.emptyMap());
+    testReclaimableRenameEntryFilter(volume, bucket, index, value, keyTable, directoryTable, true);
+  }
+
+  @ParameterizedTest
+  @MethodSource("testReclaimableFilterArguments")
+  public void testNonReclaimableRenameEntryWithFileFSO(int actualNumberOfSnapshots, int index)
+      throws IOException, RocksDBException {
+    setup(1, actualNumberOfSnapshots, index, 4, 2,
+        BucketLayout.FILE_SYSTEM_OPTIMIZED);
+    String volume = getVolumes().get(3);
+    String bucket = getBuckets().get(1);
+    index = Math.min(index, actualNumberOfSnapshots);
+    String value = UUID.randomUUID().toString();
+    Table<String, OmKeyInfo> keyTable = getMockedTable(ImmutableMap.of(value, mock(OmKeyInfo.class)));
+    Table<String, OmDirectoryInfo> directoryTable = getMockedTable(Collections.emptyMap());
+    testReclaimableRenameEntryFilter(volume, bucket, index, value, keyTable, directoryTable, index == 0);
+  }
+
+  @ParameterizedTest
+  @MethodSource("testReclaimableFilterArguments")
+  public void testNonReclaimableRenameEntryWithDirFSO(int actualNumberOfSnapshots, int index)
+      throws IOException, RocksDBException {
+    setup(1, actualNumberOfSnapshots, index, 4, 2,
+        BucketLayout.FILE_SYSTEM_OPTIMIZED);
+    String volume = getVolumes().get(3);
+    String bucket = getBuckets().get(1);
+    index = Math.min(index, actualNumberOfSnapshots);
+    String value = UUID.randomUUID().toString();
+    Table<String, OmKeyInfo> keyTable = getMockedTable(Collections.emptyMap());
+    Table<String, OmDirectoryInfo> directoryTable = getMockedTable(ImmutableMap.of(value, mock(OmDirectoryInfo.class)));
+    testReclaimableRenameEntryFilter(volume, bucket, index, value, keyTable, directoryTable, index == 0);
```

Comment:
reclaimable if key is in the key table or directory table and its the first snapshot (or there's no snapshot) in a FSO bucket.